src/infra/src/table/timed_annotations.rs

```diff
@@ -0,0 +1,481 @@
+        .one(client)
+        .await?
+        .ok_or_else(|| {
+            errors::Error::DbError(errors::DbError::KeyNotExists(format!(
+                "TimedAnnotation with ID {} not found in dashboard {}",
+                timed_annotation_id, dashboard_id
+            )))
+        })?;
+
+    // Step 3: Fetch associated panels for the annotation
+    let panels = timed_annotation_panels::Entity::find()
+        .filter(timed_annotation_panels::Column::TimedAnnotationId.eq(timed_annotation_id))
+        .all(client)
+        .await?;
+
+    // Step 4: Collect panel IDs
+    let panel_ids: Vec<String> = panels.into_iter().map(|panel| panel.panel_id).collect();
+
+    // Step 5: Map the database model (`timed_annotations::Model`) to the domain model
+    // (`TimedAnnotation`)
+    let timed_annotation = TimedAnnotation {
+        annotation_id: Some(annotation.id),
+        start_time: annotation.start_time,
+        end_time: annotation.end_time,
+        title: annotation.title,
+        text: annotation.text,
+        tags: annotation
+            .tags
+            .as_array()
+            .unwrap_or(&vec![])
+            .iter()
+            .filter_map(|v| v.as_str().map(String::from))
+            .collect(),
+        panels: panel_ids,
+    };
+
+    // Step 6: Return the result
+    Ok(timed_annotation)
+}
+
+pub async fn update(
+    dashboard_id: &str,
+    timed_annotation_id: &str,
+    timed_annotation: TimedAnnotationUpdate,
+) -> Result<(), errors::Error> {
+    // make sure only one client is writing to the database(only for sqlite)
+    let _lock = get_lock().await;
+
+    // Initialize the ORM client
+    let client = ORM_CLIENT.get_or_init(connect_to_orm).await;
+
+    // Step 1: Resolve the user-facing `dashboard_id` to the primary key
+    let dashboard_record = dashboards::Entity::find()
+        .filter(dashboards::Column::DashboardId.eq(dashboard_id))
+        .one(client)
+        .await?
+        .ok_or_else(|| {
+            errors::Error::DbError(errors::DbError::KeyNotExists(format!(
+                "Dashboard '{}' not found",
+                dashboard_id
+            )))
+        })?;
+
+    let dashboard_pk = dashboard_record.id;
+
+    // Step 2: Build the update query
+    let mut update_query = timed_annotations::Entity::update_many()
+        .filter(timed_annotations::Column::DashboardId.eq(dashboard_pk))
+        .filter(timed_annotations::Column::Id.eq(timed_annotation_id));
+
+    // Step 3: Apply updates from the `TimedAnnotationUpdate` struct
+    if let Some(start_time) = timed_annotation.start_time {
+        update_query = update_query.col_expr(
+            timed_annotations::Column::StartTime,
+            Expr::value(start_time),
+        );
+    }
+    if let Some(end_time) = timed_annotation.end_time {
+        update_query =
+            update_query.col_expr(timed_annotations::Column::EndTime, Expr::value(end_time));
+    }
```

Comment:
Related to this [other comment](https://github.com/openobserve/openobserve/pull/5680#discussion_r1925571906)

Currently the UpdateTimedAnnotation endpoint only supports 
- leaving `end_time` unchanged
- or setting `end_time` to a different time

If we want to add support for "unsetting" `end_time` (ie setting it to `None`), then we may want to consider changing the `end_time` field on the `TimedAnnoationUpdate` struct so that it has the type `Option<Option<i64>>`

Reply:
Modified to handle as per the `Option<Option<i64>>`